File: bench-langs/eisenstein_bench.py
"""Eisenstein Kernel Benchmark — Python"""
import time
import math
import random

random.seed(42)

def eisenstein_norm(a: int, b: int) -> int:
    """N(a + bw) = a^2 - a*b + b^2"""
    return a * a - a * b + b * b

def eisenstein_snap(x: float, y: float) -> tuple:
    """Snap (x,y) to nearest Eisenstein lattice point"""
    q = (2.0/3.0 * x - 1.0/3.0 * y)
    r = (2.0/3.0 * y)
    rq = round(q)
    rr = round(r)
    rs = round(-q - r)
    diff = abs(rq + rr + rs)
    if diff == 0:
        a, b = rq, rr
    elif diff == 2:
        if abs(rq - q) > abs(rr - r):
            rq = -rr - rs
        else:
            rr = -rq - rs
        a, b = rq, rr
    else:
        a, b = rq, rr
    return a, b

def eisenstein_distance(x: float, y: float, a: int, b: int) -> float:
    """Distance from (x,y) to lattice point (a, bw)"""
    px = a - 0.5 * b
    py = b * math.sqrt(3) / 2.0
    return math.sqrt((x - px)**2 + (y - py)**2)

def constraint_check(a: int, b: int, radius: float) -> bool:
    """Check if Eisenstein norm is within radius constraint"""
    norm = eisenstein_norm(a, b)
    return norm <= radius * radius

# Generate test data
N = 10_000_000
norm_inputs = [(random.randint(-1000, 1000), random.randint(-1000, 1000)) for _ in range(N)]
snap_inputs = [(random.uniform(-100, 100), random.uniform(-100, 100)) for _ in range(N)]
constraint_inputs = [(random.randint(-100, 100), random.randint(-100, 100), random.uniform(1, 50)) for _ in range(N)]

# Benchmark norm
start = time.perf_counter()
norm_sum = sum(eisenstein_norm(a, b) for a, b in norm_inputs)
norm_time = time.perf_counter() - start

# Benchmark snap
start = time.perf_counter()
snap_results = [eisenstein_snap(x, y) for x, y in snap_inputs]
snap_time = time.perf_counter() - start

# Benchmark constraint
start = time.perf_counter()
constraint_pass = sum(1 for a, b, r in constraint_inputs if constraint_check(a, b, r))
constraint_time = time.perf_counter() - start

print(f"Python Results (N={N:,}):")
print(f"  eisenstein_norm:  {norm_time:.3f}s  (sum={norm_sum})")
print(f"  eisenstein_snap:  {snap_time:.3f}s  (first={snap_results[0]})")
print(f"  constraint_check: {constraint_time:.3f}s  (pass={constraint_pass})")
print(f"  TOTAL: {norm_time + snap_time + constraint_time:.3f}s")

File: bench-langs/eisenstein_bench.c
// Eisenstein Kernel Benchmark — C
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

typedef long long i64;

static inline i64 eisenstein_norm(i64 a, i64 b) {
    return a * a - a * b + b * b;
}

typedef struct { i64 a; i64 b; } IntPair;
typedef struct { double x; double y; } DblPair;

IntPair eisenstein_snap(double x, double y) {
    double q = (2.0/3.0 * x - 1.0/3.0 * y);
    double r = (2.0/3.0 * y);
    double rq = round(q);
    double rr = round(r);
    double rs = round(-q - r);
    double diff = fabs(rq + rr + rs);
    if (diff == 2.0) {
        if (fabs(rq - q) > fabs(rr - r)) {
            rq = -rr - rs;
        } else {
            rr = -rq - rs;
        }
    }
    IntPair result = {(i64)rq, (i64)rr};
    return result;
}

double eisenstein_distance(double x, double y, i64 a, i64 b) {
    double px = a - 0.5 * b;
    double py = b * 0.8660254037844386; // sqrt(3)/2
    double dx = x - px;
    double dy = y - py;
    return sqrt(dx*dx + dy*dy);
}

int constraint_check(i64 a, i64 b, double radius) {
    return eisenstein_norm(a, b) <= (i64)(radius * radius);
}

#define N 10000000

int main() {
    srand(42);

    // Generate data
    i64 *norm_a = malloc(N * sizeof(i64));
    i64 *norm_b = malloc(N * sizeof(i64));
    double *snap_x = malloc(N * sizeof(double));
    double *snap_y = malloc(N * sizeof(double));
    i64 *con_a = malloc(N * sizeof(i64));
    i64 *con_b = malloc(N * sizeof(i64));
    double *con_r = malloc(N * sizeof(double));

    for (int i = 0; i < N; i++) {
        norm_a[i] = (rand() % 2001) - 1000;
        norm_b[i] = (rand() % 2001) - 1000;
        snap_x[i] = ((double)rand() / RAND_MAX) * 200.0 - 100.0;
        snap_y[i] = ((double)rand() / RAND_MAX) * 200.0 - 100.0;
        con_a[i] = (rand() % 201) - 100;
        con_b[i] = (rand() % 201) - 100;
        con_r[i] = ((double)rand() / RAND_MAX) * 49.0 + 1.0;
    }

    struct timespec t0, t1;
    i64 norm_sum = 0;
    IntPair snap_first = {0, 0};
    i64 con_pass = 0;

    // Benchmark norm
    clock_gettime(CLOCK_MONOTONIC, &t0);
    for (int i = 0; i < N; i++) {
        norm_sum += eisenstein_norm(norm_a[i], norm_b[i]);
    }
    clock_gettime(CLOCK_MONOTONIC, &t1);
    double norm_time = (t1.tv_sec - t0.tv_sec) + (t1.tv_nsec - t0.tv_nsec) / 1e9;

    // Benchmark snap
    clock_gettime(CLOCK_MONOTONIC, &t0);
    for (int i = 0; i < N; i++) {
        IntPair s = eisenstein_snap(snap_x[i], snap_y[i]);
        if (i == 0) snap_first = s;
    }
    clock_gettime(CLOCK_MONOTONIC, &t1);
    double snap_time = (t1.tv_sec - t0.tv_sec) + (t1.tv_nsec - t0.tv_nsec) / 1e9;

    // Benchmark constraint
    clock_gettime(CLOCK_MONOTONIC, &t0);
    for (int i = 0; i < N; i++) {
        con_pass += constraint_check(con_a[i], con_b[i], con_r[i]);
    }
    clock_gettime(CLOCK_MONOTONIC, &t1);
    double con_time = (t1.tv_sec - t0.tv_sec) + (t1.tv_nsec - t0.tv_nsec) / 1e9;

    printf("C Results (N=%d):\n", N);
    printf("  eisenstein_norm:  %.3fs  (sum=%lld)\n", norm_time, norm_sum);
    printf("  eisenstein_snap:  %.3fs  (first=(%lld,%lld))\n", snap_time, snap_first.a, snap_first.b);
    printf("  constraint_check: %.3fs  (pass=%lld)\n", con_time, con_pass);
    printf("  TOTAL: %.3fs\n", norm_time + snap_time + con_time);

    free(norm_a); free(norm_b); free(snap_x); free(snap_y);
    free(con_a); free(con_b); free(con_r);
    return 0;
}
